src/westmere/ip4.h

```diff
@@ -0,0 +1,210 @@
+  {2, 1, 4, 128, 8, 7, 10, 128, 0, 0, 128, 128, 6, 6, 128, 128},
+  {2, 1, 4, 128, 8, 7, 11, 10, 0, 0, 128, 128, 6, 6, 128, 10},
+  {2, 1, 4, 128, 8, 7, 12, 11, 0, 0, 128, 128, 6, 6, 10, 10},
+  {2, 1, 5, 4, 7, 128, 9, 128, 0, 0, 128, 4, 128, 128, 128, 128},
+  {2, 1, 5, 4, 7, 128, 10, 9, 0, 0, 128, 4, 128, 128, 128, 9},
+  {2, 1, 5, 4, 7, 128, 11, 10, 0, 0, 128, 4, 128, 128, 9, 9},
+  {2, 1, 5, 4, 8, 7, 10, 128, 0, 0, 128, 4, 128, 7, 128, 128},
+  {2, 1, 5, 4, 8, 7, 11, 10, 0, 0, 128, 4, 128, 7, 128, 10},
+  {2, 1, 5, 4, 8, 7, 12, 11, 0, 0, 128, 4, 128, 7, 10, 10},
+  {2, 1, 5, 4, 9, 8, 11, 128, 0, 0, 128, 4, 7, 7, 128, 128},
+  {2, 1, 5, 4, 9, 8, 12, 11, 0, 0, 128, 4, 7, 7, 128, 11},
+  {2, 1, 5, 4, 9, 8, 13, 12, 0, 0, 128, 4, 7, 7, 11, 11},
+  {2, 1, 6, 5, 8, 128, 10, 128, 0, 0, 4, 4, 128, 128, 128, 128},
+  {2, 1, 6, 5, 8, 128, 11, 10, 0, 0, 4, 4, 128, 128, 128, 10},
+  {2, 1, 6, 5, 8, 128, 12, 11, 0, 0, 4, 4, 128, 128, 10, 10},
+  {2, 1, 6, 5, 9, 8, 11, 128, 0, 0, 4, 4, 128, 8, 128, 128},
+  {2, 1, 6, 5, 9, 8, 12, 11, 0, 0, 4, 4, 128, 8, 128, 11},
+  {2, 1, 6, 5, 9, 8, 13, 12, 0, 0, 4, 4, 128, 8, 11, 11},
+  {2, 1, 6, 5, 10, 9, 12, 128, 0, 0, 4, 4, 8, 8, 128, 128},
+  {2, 1, 6, 5, 10, 9, 13, 12, 0, 0, 4, 4, 8, 8, 128, 12},
+  {2, 1, 6, 5, 10, 9, 14, 13, 0, 0, 4, 4, 8, 8, 12, 12},
+};
+
+static int sse_inet_aton(const char* ipv4_string, const size_t ipv4_string_length, uint8_t * destination) {
+  //const size_t n = ipv4_string_length;
+  // This function always reads 16 bytes. With AVX-512 we can do a mask
+  // load, but it is not generally available with SSE 4.1.
+  const __m128i input = _mm_loadu_si128((const __m128i *)ipv4_string);
+  if (ipv4_string_length > 15) {
+    return 0;
+  }
+  // locate dots
+  uint16_t dotmask;
+  {
+    const __m128i dot = _mm_set1_epi8('.');
+    const __m128i t0 = _mm_cmpeq_epi8(input, dot);
+    dotmask = _mm_movemask_epi8(t0);
+    uint16_t mask = 1 << ipv4_string_length;
+    dotmask &= mask - 1;
+    dotmask |= mask;
+  }
+
+  // build a hashcode
+  const uint8_t hashcode = ((6639 * dotmask) >> 13);
+
+  // grab the index of the shuffle mask
+  const uint8_t id = patterns_id[hashcode];
+  if (id >= 81) {
+    return 0;
+  }
+  const uint8_t *pat = &patterns[id][0];
+  const __m128i pattern = _mm_loadu_si128((const __m128i *)pat);
+  // The value of the shuffle mask at a specific index points at the last digit,
+  // we check that it matches the length of the input.
+  const __m128i ascii0 = _mm_set1_epi8('0');
+  const __m128i t0 = input;
+  __m128i t1 = _mm_shuffle_epi8(t0, pattern);
+  // check that leading digits of 2- 3- numbers are not zeros.
+  {
+    const __m128i eq0 = _mm_cmpeq_epi8(t1, ascii0);
+    if (!_mm_testz_si128(eq0, _mm_set_epi8(-1, 0, -1, 0, -1, 0, -1, 0,
+                                           0, 0, 0, 0, 0, 0, 0, 0))) {
+      return 0;
+    }
+  }
+  // replace null values with '0'
+  __m128i t1b = _mm_blendv_epi8(t1, ascii0, pattern);
```

Comment:
I'm wondering if this can be simplified. Next to the pattern, we could use an AND-mask and save that with the shuffle pattern? So, `_mm_sub_epi8` then and with the mask (I'm assuming we need it to zero out bytes that weren't actually set in the textual presentation) and then simply do a `_mm_cmpgt_epi8(0x09)` (instead of `_mm_max_epu8` + `_mm_cmpeq_epi8` + `_mm_test_all_ones`). This may be a completely bogus idea (I should really study the algorithm better, I'll do so today or tomorrow, but it's something that popped into my head...)

Reply:
The "check that everything was in the range '0' to '9'" check could be combined with with the overflow check.

`_mm_maddubs_epi16` sign extends, so the trick is to set bad chars to `0x80` or some other value (or range) that would work. Then use `_mm_adds_epu16()` to combine the high and low 64-bit lanes. 

Note: the intrinsic guide shows `_mm_test_all_ones()` as a sequence unlike all the other `PTEST` mnemonics.

Reply:
@k0ekk0ek An alternative, if you don't like `_mm_max_epu8` is to do...

```C++
    const __m128i t2z = _mm_add_epi8(t2,_mm_set1_epi8(-128) );
    const __m128i c9 = _mm_set1_epi8('9' - '0' - 128);
    const __m128i t2me = _mm_cmpgt_epi8(t2z, c9);
    if (!_mm_test_all_zeros(t2me, t2me)) {
      return 0;
    }
```

At least with GCC, this results in exactly the same instruction count and no benefit performance-wise.

> then simply do a _mm_cmpgt_epi8(0x09) (instead of _mm_max_epu8 + _mm_cmpeq_epi8 + _mm_test_all_ones)

Are you taking into account that `_mm_cmpgt_epi8` is a signed comparison? This means that 0x80 is smaller than 0x09 (among other things).

Reply:
I spent some time on this code and there is not a whole lot to save. If you omit the constants and the if clause, we are talking about ~2 instructions. If we could combine it with some other check, like @aqrit, that might save us quite a bit, e.g., by taking out one branch. However, it is not obvious how to do it safely.

However, I am going to have to write a routine that checks that we only have dots and digits, and if I am going to do that anyhow, I will be able to save on this routine here (we just need to check that there is no '.' since the only possibilities are digits and dots).

Reply:
Not too worried about `_mm_max_epu8` per se. It's more that we'll need the length (or if there's another way, that's fine too) to verify there's a delimiter right after the address (after #64 is merged). I figured picking up a __m128i (mask) to do an `_mm_and_si128` to zero out non-digits (as opposed to `_mm_blendv_epi8`), move the check forward and base it on a bitmask instead, we may even be a little faster(?) The bitmask could also be used to count trailing zeroes (loading that from a table is another possibility). My thinking was: if there's any bit set before the delimiter, that is not set in the `dotmask` or `digit_mask`, we'd have an invalid input. Maybe even combine multiple checks (either by doing them sooner or by delaying and doing them later) that would save even more (as @aqrit suggested). And you're right, I did not take into account `_mm_cmpgt_epi8` is signed :sweat_smile:. Only after toying with the code yesterday, did I find out.

Reply:
I just pushed a commit that provides two functions, one with a length parameter and one without (`sse_inet_aton_16`). The new code does not do the `_mm_max_epu8` thing nor does it do blending because we know that we only have digits or dots to check (or else, the length will not match). So you save some work there, but you pay for the length computation. It a net loss in throughput unfortunately because of the long chain of dependencies...